### PR TITLE
Excel & CSV query runner

### PR DESCRIPTION
**Pull request by deecay**
This PR adds Excel and CSV as possible datasource.

* Datasource Name is the only configuration for both datasource.
* Query must be in YAML format.
* Specify path of excel/csv file, either local path or URI, for `url` parameter.
* You may pass other parameters to the pandas excel/csv parsing function.

- Most of the code is taken from #2078, thanks to @dersphere.

**Online Excel [data file](https://www.unicef.org/sowc2012/pdfs/U5MR-rank_FINAL.xls) with parameters**
image

- Reference for parameters: [CSV](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) and [Excel](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html).

**Same data without parameters** (table is broken)
* `https://www.unicef.org/sowc2012/pdfs/U5MR-rank_FINAL.xls`
image

* TODO:

 - [ ] db-logo

### CODE REVIEW

**Comment by deecay** [comment]
Need db-logo for Excel...

**Comment by deecay** [comment]
@jezdez, any opinions? Maybe resort to `requirements_excel_ds.txt`?

**Comment by jezdez** [comment]
Yeah, installing Pandas and numpy is quite a lot for the purpose of reading and parsing Excel and CSV files alone. That's an additional ~23 MB (unpacked whl files on Linux for Python 2.7) for the Docker image and a higher maintenance burden given the high profile of Pandas and numpy for a relatively small, albeit useful feature.

Options forward:

- find a different library to read and parse CSV/Excel files
- ship the query runner as a separate package (#2921) and document how users can install it on demand (my preference)
- accept the extra dependency

A positive point is that both Pandas and numpy are available as precompiled whl files, so it's literally just a matter of downloading it.

**Comment by denisov-vlad** [comment]
Pandas is must have for python query runner.

Maybe it's heavy but does not require additional libraries which should be installed via apt like other datasources like mssql.

**Comment by jezdez** [comment]
> Pandas is must have for python query runner.
> 
> Maybe it's heavy but does not require additional libraries which should be installed via apt like other datasources like mssql.

That may or may not be so, it's out of scope of this pull request review though.

**Comment by deecay** [comment]
Option 2 is acceptable for me too. I will wait until #2921 is done.

`xlrd` will be the candidate for Option 1, but this involves decent amount of 'reinventing the wheel' to get the nice features that Pandas have (skiprows, usecols, etc). Without these features, a "perfectly formatted" excel will be required, which can be very rare as I have displayed in the examples.

**Comment by arikfr** [comment]
> Option 2 is acceptable for me too. I will wait until #2921 is done.

We don't have to extract all the query runners for you to be able to do this. But we will probably need to extract the query runners base classes and helper methods to their own package, so they can be used "externally". @jezdez , am I correct here?

> xlrd will be the candidate for Option 1, but this involves decent amount of 'reinventing the wheel' to get the nice features that Pandas have (skiprows, usecols, etc). Without these features, a "perfectly formatted" excel will be required, which can be very rare as I have displayed in the examples.

I missed the added functionality. That's actually nice :) I would suggest a different "query syntax" though, let's use YAML here, so the query becomes:

```yaml
url: https://www.unicef.org/sowc2012/pdfs/U5MR-rank_FINAL.xls
names:
  - Country
  - Mortality
  - Rank
usecols: [0, 1, 2]
skiprows: 7
skipfooter: 2
```

**Comment by deecay** [comment]
Done with the yaml part.
image

**Comment by deecay** [comment]
Is this okay to merge now, since we have both numpy and pandas in our standard install?

**Comment by machzqcq** [comment]
> Is this okay to merge now, since we have both numpy and pandas in our standard install?

Can someone confirmed this is merged, reading excel and pandas df as data source will help a lot

**Comment by dersphere** [comment]
Would be great to have my initial code somehow integrated (thanks @deecay) :)

**Comment by xsvfat** [comment]
Would be great to get this merged!

**Review by jezdez — COMMENTED** [review]
I'm sorry this never went anywhere folks.

**Comment by deecay** [comment]
> I'm sorry this never went anywhere folks.

Hi @jezdez, what is blocking this? Or should I close this for some reason?

**Review by arikfr — APPROVED** [review]
Good news: we can merge this! @deecay can you please confirm this still works on latest master?

And sorry it took so long.

**Review by arikfr — COMMENTED** [review]
Actually there are few things missing:

1. Logos for each one of them. 
2. Register them so they loaded by default. 
3. Add the local urls filtering (let me know if you need a pointer for what I mean). 

Thanks!

**Comment by deecay** [comment]
1. Logo for Excel added. (csv logo was already there)
2. Updated init file to be enabled by default

Could you point to the local urls filtering?

**Comment by susodapop** [comment]
> Could you point to the local urls filtering?

I _think_ Arik is referring to [this](https://github.com/getredash/redash/pull/3805#discussion_r299353218) which was discussed when implementing the JSON data source.

File: redash/query_runner/excel.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        ua = ""
+        args = {}
+        try:
+            args = yaml.safe_load(query)
+            path = args['url']
+            args.pop('url', None)
+            ua = args['user-agent']
+            args.pop('user-agent', None)
+
+            if is_private_address(path) and settings.ENFORCE_PRIVATE_ADDRESS_BLOCK:
```
Comment (deecay) [inline redash/query_runner/excel.py:52]: @susodapop , I _guess_ you're right. Done.

**Comment by arikfr** [comment]
So it took over 3 years but it's merged now. Thank you for the the continued contribution and immense patience 😅

**Comment by dersphere** [comment]
😃

**Comment by ainsofs** [comment]
CSV and Excel do not appear on my list of data sources. Is there something I need to do to make that happen? I am using image redash/redash:8.0.0.b32245 . I also tried image version 10 beta with the same result

**Comment by susodapop** [comment]
@ainsofs You need to run the tip of `master` since this PR isn't merged into the V10 branch (yet). It will be before we release.

**Pull request merged**